Add support for formatted lock file names

pbjc opened the pull request:
Partly addresses #69

mariusvniekerk commented on conda_lock/conda_lock.py:
```diff
@@ -534,6 +554,11 @@ def main():
     help="path to a conda environment specification(s)",
 )
+@click.option(
+    "--filename-format",
```
I'd rename this to filename-template

bollwyvl commented on conda_lock/conda_lock.py:
```diff
@@ -534,6 +555,11 @@ def main():
     help="path to a conda environment specification(s)",
 )
+@click.option(
+    "--filename-template",
```
Perhaps `-o` and `--output` might be less surprising?

pbjc replied:
I have no preference, I'll leave this up to the maintainers

bollwyvl commented on conda_lock/conda_lock.py:
```diff
@@ -36,6 +36,7 @@
 
 CONDA_PKGS_DIRS = None
 DEFAULT_PLATFORMS = ["osx-64", "linux-64", "win-64"]
+FILENAME_TOKENS = frozenset(["platform", "dev-dependencies", "spec-hash"])
```
Perhaps expanding these to be a dict, e.g. 
```py
FILENAME_TOKENS = {
    "platform": "the conda-subdir",
    "dev-dependencies": "..."
    "spec-hash": "..."
}
```
And these can be included in the `help`.

pbjc replied:
Turns out it's kind of tricky to have newlines in the help for an argument, and also to have a formatted string as your docstring, so I just added a hardcoded description of the tokens to the command help.

bollwyvl commented:
I really,  actually meant to get around to doing this... hooray @pbjc for beating me to it :tada: :t-rex: !

I dropped a few comments above. In addition to the ones already loaded up, perhaps:
- the version of `conda-lock`: `{version}` -> `0.7.4`
- the ISO8601 timestamp, e.g. `{data}` -> `2021-01-01T000000Z` (colons make windows sad)
- the name/version of the tool that did the solving, e.g. `{solver}-{solver_version}` -> `mamba-0.7.7`

pbjc commented:
> I really, actually meant to get around to doing this... hooray @pbjc for beating me to it 🎉 🦖 !
> 
> I dropped a few comments above. In addition to the ones already loaded up, perhaps:
> 
> * the version of `conda-lock`: `{version}` -> `0.7.4`
> * the ISO8601 timestamp, e.g. `{data}` -> `2021-01-01T000000Z` (colons make windows sad)
> * the name/version of the tool that did the solving, e.g. `{solver}-{solver_version}` -> `mamba-0.7.7`

Added version and timestamp! Solver and solver version are trickier, it can be done but I think it would be cleaner for ensureconda to just add support for this. https://github.com/conda-incubator/ensureconda/issues/13

bollwyvl commented:
Fantastic! Thank you so much!
